packages/devextreme/js/__internal/ui/chat/messagebox.ts

```diff
@@ -126,18 +159,52 @@ class MessageBox extends DOMComponent<MessageBox, Properties> {
     });
   }
 
+  _onInputTriggerTypingEventsHandler(e: InputEvent): void {
+    if (this._isValuableTextEntered()) {
+      this._throttledTriggerTypingStartAction?.({ event: e.event });
+    }
+  }
+
   _createMessageSendAction(): void {
     this._messageSendAction = this._createActionByOption(
       'onMessageSend',
       { excludeValidators: ['disabled', 'readOnly'] },
     );
   }
 
+  _createTypingStartAction(): void {
+    this._typingStartAction = this._createActionByOption(
+      'onTypingStart',
+      { excludeValidators: ['disabled', 'readOnly'] },
+    );
+  }
+
+  _createTypingEndAction(): void {
+    this._typingEndAction = this._createActionByOption(
+      'onTypingEnd',
+      { excludeValidators: ['disabled', 'readOnly'] },
+    );
+  }
+
+  _triggerTypingStartAction(e: InputEvent): void {
+    clearTimeout(this._typingEndEventTimeout);
+
+    this._typingEndEventTimeout = setTimeout(() => {
+      this._typingEndAction?.();
+    }, TYPING_END_DELAY);
+
+    this._typingStartAction?.({ event: e.event });
+  }
+
   _sendHandler(e: ClickEvent | EnterKeyEvent): void {
     if (!this._isValuableTextEntered()) {
       return;
     }
 
+    clearTimeout(this._typingEndEventTimeout);
```

Comment:
Can we extract part `clearTimeout(this._typingEndEventTimeout)` to separate function `clearTypingEndTimer()' to avoid duplication?

Reply:
[9345dcc44d2d5822ffdceb0dca58e69e99d62bc9]